src/client/main.js

```diff
@@ -67,7 +67,12 @@ export function define(cell) {
 function noop() {}
 
 function clear(root) {
-  for (const v of root._nodes) v.remove();
+  delete root._inspectorState;
+  for (const v of root._nodes) {
+    // If this was an inspector, note whether it is expanded.
+    if (v.firstChild?.classList.contains("observablehq--expanded")) root._inspectorExpanded = true;
```

Comment:
So, if any of the displayed items was expanded, all of them will be expanded on the next display? (Unlike in notebooks, code blocks in Framework can call `display` multiple times.)

I prefer that either we track the index of the expanded items and restore them consistently, or we special-case the expanded state of the first item, or we special-case the expanded state for the display of a singular item. I don’t think we really need to handle restoring the expanded state of multiple displayed items, but I don’t want to accidentally expand a lot of displayed items, either.

Reply:
I've now changed the logic to store a boolean for each node; however I don't think there is any way to absolutely guarantee constancy if the code block is playing tricks like this:

~~~
```js
if (Math.random() > 0.5) display(foo);
display(bar);
```
~~~

similarly:

~~~
```js
display(Math.random() > 0.5 ? foo : bar);
```
~~~

you will expand foo, refresh, and see bar expanded. I think it's OK, though.

Reply:
> represent inspector state consistently by initializing it when the root node is attached

I'm not sure I understand this comment, since the state varies with user interaction. Either we have to listen to changes, or we read the state when we "unmount" the node to make room for the new versions.

Reply:
All I meant here is that we should do something like

```js
root._inspectorExpanded = [];
```

around here when the `root` is first bound:

https://github.com/observablehq/framework/blob/e123e7a85d61c83557a555fd005e0246037cbdd3/src/client/main.js#L31